### PR TITLE
feat(dap): send Disconnect if Terminated event received

### PR DESCRIPTION
**Pull request by filipdutescu**
Send a `Disconnect` DAP request if the `Terminated` event is received. According to the specification, if the debugging session was started by as `launch`, the debuggee should be terminated alongside the session. If instead the session was started as `attach`, it should not be disposed of.

This default behaviour can be overriden if the `supportTerminateDebuggee` capability is supported by the adapter, through the `Disconnect` request `terminateDebuggee` argument, as described in
[the specification][discon-spec].

This also implies saving the starting command for a debug sessions, in order to decide which behaviour should be used, as well as validating the capabilities of the adapter, in order to decide what the disconnect should do.

An additional change made is handling of the `Exited` event, showing a message if the exit code is different than `0`, for the user to be aware off the termination failure.

[discon-spec]: https://microsoft.github.io/debug-adapter-protocol/specification#Requests_Disconnect

Closes: #4674
Signed-off-by: Filip Dutescu <filip.dutescu@gmail.com>

### CODE REVIEW

File: helix-view/src/handlers/dap.rs
Comment (filipdutescu) [inline helix-view/src/handlers/dap.rs:313]: restart should be done with the [`Restart Request`](https://microsoft.github.io/debug-adapter-protocol/specification#Requests_Restart) :thinking:

**Review by filipdutescu — COMMENTED** [review]
just realised the restart implementation I made is not correct according to the spec. will redo it

File: helix-view/src/handlers/dap.rs
Comment (filipdutescu) [reply helix-view/src/handlers/dap.rs:313]: I was incorrect in my assumption. the request should be used to _*trigger*_ a restart, rather than "create one". In any case, this should be further elaborated, most likely in a different PR, handling restarting as a whole

**Comment by filipdutescu** [comment]
I have re-read the spec and I believe this part of the restart sequence to be correct. What needs to be done in addition is to expose a restart request, if supported, to the user, same as with launch, terminate etc.

Cc: @pascalkuthe could we move it to waiting on review? Many thanks!

File: helix-dap/src/client.rs
Comment (archseer) [inline helix-dap/src/client.rs:49]: Remove `None` and instead use `Option<StartingRequest>`

Comment (archseer) [reply helix-dap/src/client.rs:49]: Also this seems better named as connection type?

**Review by archseer — CHANGES_REQUESTED** [review]

File: helix-dap/src/client.rs
Comment (archseer) [inline helix-dap/src/client.rs:89]: This should also be renamed to connection_type now

File: helix-dap/src/client.rs
Comment (archseer) [inline helix-dap/src/client.rs:218]: Same here

File: helix-view/src/handlers/dap.rs
Comment (archseer) [inline helix-view/src/handlers/dap.rs:296]: None should come after Some

File: helix-view/src/handlers/dap.rs
Comment (archseer) [inline helix-view/src/handlers/dap.rs:293]: Rewrite this so it's 

```
if let Err(ere) = result {
   set_error(...)
   return;
}
```

That way the error exits early and the regular path isn't indented inside an if

File: helix-view/src/handlers/dap.rs
Comment (archseer) [inline helix-view/src/handlers/dap.rs:294]: Similar to the other PR you want to match on the argument so we avoid an unwrap

**Review by archseer — CHANGES_REQUESTED** [review]

**Comment by filipdutescu** [comment]
@archseer implemented your comments, thanks very much for the patience and advice, I have learnt a lot and hopefully improved the way I write Rust. I also rebased the PR and squashed the commits, to make it a bit cleaner, since there are not that many modifications (hopefully) as to make it a chore to revisit the changes.

**Review by archseer — APPROVED** [review]
Sorry it took me a while to get back to this! LGTM!

**Comment by filipdutescu** [comment]
No worries, thanks for reviewing and merging it!

**Pull request merged**
